### PR TITLE
add custom field source

### PR DESCRIPTION
**Pull request by tylerflex**
Incorporates changes from #447 with the new data refactor in mind.

Needs backend implementation.

### CODE REVIEW

**Comment by momchil-flex** [comment]
Looks good, I think some things may pop up while doing the backend implementation, but for now I just have these comments:

- We'll see how this goes, but we may not have to restrict to single frequency. We have already introduced multi-frequency (boradband) sources for ModeSource and GaussianBeam and the same approach should work here (assuming the freuqency dependence of the fields is fairly smooth).
- In the previous PR discussion, I think we agreed on something like, recenter the field dataset to the source center? There are a few intricacies here though and I think we should have some validators:
  - How exactly do we do the recentering. Say that the field data is defined on Yee grid locations in the tangential direction (as it would come from a FieldMonitor for example). Say the monitor is z-normal. For x and y, do we take something like the smallest and the largest coordinate value among all field component as the xmin/xmax, ymin/ymax of the entire data, and then shift the data such that (xmin + xmax) / 2 is at the source center? This can be done on the backend, but should probably be clarified for the user.
  - Actually what happens if a `FieldData` is provided as the dataset? Does the validation error out? Does it initialize a `FieldDataset` instead of `FieldData`? Or does it actually initialize a `FieldData` (and even after IO?) We should have the easiest possible way to do something like `custom_source = td.CustomSource(field_dataset=sim_data["field_monitor"])`, i.e. use the data from one simulation's monitor for a source in another simulation. If this produces a custom source with a `FieldData` as the dataset, the best thing to do regarding recentering would be to shift the data by the difference between `source.center` and `FieldData.monitor.center`.
  - In any case should we check that each provided field component fully covers (after recentering) the source size in the tangential directions? I think that makes sense, apart from case where you may want to make the source size `td.inf` for convenience, but maybe you'll just have to not do that?
  - Should we check that the fields have a single coordinate along the normal direction? Or leave this to just also be handled by the re-centering? (i.e. if multiple coordinates the fields at the center of the data along the normal direction will be injected)

**Comment by momchil-flex** [comment]
> > * We'll see how this goes, but we may not have to restrict to single frequency. We have already introduced multi-frequency (boradband) sources for ModeSource and GaussianBeam and the same approach should work here (assuming the freuqency dependence of the fields is fairly smooth).
> 
> Sounds good
> 
> > * In the previous PR discussion, I think we agreed on something like, recenter the field dataset to the source center? There are a few intricacies here though and I think we should have some validators:
> 
> Yea, I think we decided that a (0,0,0) in the dataset corresponds to the `source.center`. Discussion [here](https://github.com/flexcompute/tidy3d/pull/447).

I don't think that's what we decided there (see below).

> > * How exactly do we do the recentering. Say that the field data is defined on Yee grid locations in the tangential direction (as it would come from a FieldMonitor for example). Say the monitor is z-normal. For x and y, do we take something like the smallest and the largest coordinate value among all field component as the xmin/xmax, ymin/ymax of the entire data, and then shift the data such that (xmin + xmax) / 2 is at the source center? This can be done on the backend, but should probably be clarified for the user.
> 
> I think if we just specify that x=0, y=0 in the data corresponds to the source.center, this removes all ambiguity right? Using the min and max of the data to define a center is another way, but seems harder to use at first glance to me.

So these two approaches are certainly not equivalent. My suggestion of centering on the data coordinates center (which to my understanding was what we agreed on in the previous PR discussion) makes it straightforward to take fields recorded *anywhere* and use them as a source in a different location. Note that e.g. in Tidy3D, FieldData coordinates origin is not the monitor center, it is the simulation center - so (0, 0, 0) is likely not covered by the field data at all, unless the monitor is centered at (0, 0, 0). Specific example: say I have a monitor at (1, 1, 1) in one simulation and I want to use it as a source at (-1, -1, -1) in another. To do this directly, we need to shift the data coordinates by the difference between source center and monitor center. If we don't have access to the monitor center (e.g. the data is not coming from a Tidy3D simulation but just some external data), we need to try to reconstruct it from the data itself, which is what I was kinda describing.

That said, we *could* do what you are proposing: assume the data coordinate (0, 0, 0) sits on the source center, and leave it up to the user to make sure this is done properly. We can also handle it internally if the data is coming from a FieldData: when initializing the CustomSource, re-center the data by the `FieldData.monitor` center already. In fact I think this approach does make more sense. Either way we need to make this very clear to the user...

**Comment by tylerflex** [comment]
I just added a `.from_field_data()` classmethod constructor in [this commit](https://github.com/flexcompute/tidy3d/pull/550/commits/c18d331716f001b5b222b40474b9922e7845e7c3)

**Comment by momchil-flex** [comment]
> I just added a `.from_field_data()` classmethod constructor in [this commit](https://github.com/flexcompute/tidy3d/pull/550/commits/c18d331716f001b5b222b40474b9922e7845e7c3). Note that it adds a dependency (through the type annotation) where `source.py` imports from `monitor_data.py`. I think this should be ok though?

Ehmmmm... I thought that the whole point of separating the Dataset and Data types was to avoid that. This sounds bad to me in view of future circular dependencies again. My impression was that `MonitorData` should not be imported in any components.

I think we should just reverse this. `MonitorData` should have a `to_custom_source` method.

File: tidy3d/components/source.py [hunk 1]
```diff
@@ -509,6 +500,114 @@ def _warn_if_large_number_of_freqs(cls, val):
...
+        """Assert only one frequency supplied and it's in source time range."""
+        if val is None:
+            return val
+        source_time = get_value(key="source_time", values=values)
+        fmin, fmax = source_time.frequency_range()
+        for name, scalar_field in val.field_components.items():
+            freqs = scalar_field.f
+            if len(freqs) != 1:
+                raise SetupError(
+                    f"`field_data.{name}` must have a single frequency, "
```
Comment (tylerflex) [inline tidy3d/components/source.py:569]: fix single quotes in these raise statements, (this and below) (sorry about that)

File: tidy3d/components/validators.py [hunk 2]
```diff
@@ -38,6 +40,14 @@
 """
 
 
+def get_value(key: str, values: dict) -> Any:
+    """Grab value from values dictionary. If not present, raise an error before continuing."""
+    val = values.get(key)
+    if val is None:
+        raise ValidationError(f"value {key} not defined, must be present to validate.")
```
Comment (tylerflex) [inline tidy3d/components/validators.py:47]: add single quotes

**Pull request merged**